### PR TITLE
GEODE-8605: Alter Gateway Sender command

### PR DESCRIPTION
**Pull request by mivanac**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [*] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Review by davebarnes97 — CHANGES_REQUESTED** [review]
These requested changes apply only to documentation. This review does not address technical correctness of the PR.

FILE: CliStrings.java:

Use consistent capitalization:

GatewaySender when referring to an actual code object (such as a class).
gateway sender (not Gateway Sender) in most other cases.

For example,
 "ID of the Gateway Sender" should be "ID of the gateway sender".
Similarly,


Please rephrase:
 "Group(s) of members on which to alter option of the Gateway Sender."
 should be:
 "Group(s) of members on which to alter the gateway sender option."

 "Name/Id of the member on which to alter option of the Gateway Sender."
 should be:
 "Name/Id of the member on which to alter the gateway sender option."

"The alert threshold for entries in a GatewaySender's queue."
should be:
"The alert threshold for entries in a gateway sender's queue."

"The batch size for the GatewaySender."
should be:
"The batch size for the gateway sender."

"The batch time interval for the GatewaySender."
should be:
"The batch time interval for the gateway sender."

"The list of fully qualified class names of GatewayEventFilters (separated by comma) to be associated with the GatewaySender. This serves as a callback for users to filter out even\
ts before dispatching to remote distributed system. e.g gateway-event-filter=com.user.filters.MyFilter1,com.user.filters.MyFilters2"
should be:
"The list of fully qualified class names of GatewayEventFilters (separated by commas) to be associated with the GatewaySender. This serves as a callback for users to filter out eve\
nts before dispatching to the remote distributed system. E.g gateway-event-filter=com.user.filters.MyFilter1,com.user.filters.MyFilters2"


"The fully qualified class name of GatewayTransportFilter to be added to the GatewaySender."
should be:
"The fully qualified class name of a GatewayTransportFilter to be added to the GatewaySender."

FILE: alter.html.md.erb

"Modifies attributes of an gateway-sender."
should be
"Modifies attributes of a gateway-sender."

"Name of the group(s) whose members's runtime configuration is to be altered."
should be:
"Name of the group(s) whose members' runtime configurations are to be altered."

Two occurrences of this one, for --alert-threshold and --batch-time-interval:
"Maximum number of milliseconds that..."
should be:
"Maximum time, in milliseconds, that..."

FILE: AlterGatewaySenderCommand.java

"Can not find an gateway sender with id..."
should be:
"Cannot find a gateway sender with id..."

Two occurrences of this one:
"Alter Gateway Sender cannot be performed for..."
should be:
"alter-gateway-sender cannot be performed for..."

FILE: AlterGatewaySenderFunction.java

"Can not find existing gateway sender"
should be:
"Cannot find existing gateway sender"

FILE: all test sources, multiple occurrences:

"Can not" should be "Cannot"

"an gateway sender" should be "a gateway sender"

"Alter Gateway Sender cannot be performed"
should be:
"alter-gateway-sender cannot be performed"

**Comment by davebarnes97** [comment]
@mivanac Thanks for incorporating my suggestions. Three were incomplete, though, so I'd like to request tweaks.
Two occurrences of "in milliseconds" should have a comma afterwards: "in milliseconds,".
One occurrence of "whose members runtime configurations" should have an apostrophy after " members " to get the plural possessive " members' ". Final form should be "whose members' runtime configurations" .

**Review by davebarnes97 — APPROVED** [review]
I approve the docs. Thanks for your contribution, @mivanac.
(Please do not merge until you've also received a code review.)

**Review by kohlmu-pivotal — DISMISSED** [review]
@mivanac Thank you for this submission.

Whilst I can understand that there are certain values that one would like to change at runtime, I think there needs to be a little more thought applied to this feature change. As I don't think we have considered all the implications of this change within the system.

Could you possibly submit an RFC detailing the feature. This way this allows for more of the community to think about this change.

**Comment by nabarunnag** [comment]
Will wait assign team to review, in case this is decided to have an RFC, and if more changes are requested. Will review it in case otherwise.

**Comment by kohlmu-pivotal** [comment]
@nabarunnag thank you for the response. I would prefer if this feature change goes through an RFC process, as I believe that there might be more fields that we would want to change at runtime, or scenarios that we will have to consider.

**Comment by mivanac** [comment]
RFC created https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=165223754

**Comment by kohlmu-pivotal** [comment]
@mivanac I was just wondering if this PR is still a work in progress or if it is ready to be reviewed?
If still in progress, could you please change the status to "Draft".

**Comment by mivanac** [comment]
This PR is ready to be reviewed.

**Comment by mivanac** [comment]
Hi, is there any update with review. If there is no comments can we remove CR.

File: geode-core/src/main/resources/org/apache/geode/internal/sanctioned-geode-core-serializables.txt
Comment (bschuchardt) [inline geode-core/src/main/resources/org/apache/geode/internal/sanctioned-geode-core-serializables.txt:119]: What are the circumstances where CacheConfig$GatewaySender is serialized?  Will this cause backward-compatibility issues during a rolling upgrade?

Please add a serialversionuid field to GatewaySender.

**Review by bschuchardt — CHANGES_REQUESTED** [review]

**Comment by jinmeiliao** [comment]
So looks like you added a few options to the `alter gateway-sender` command. Do these options apply in the `create gateway-sender` as well? Would these options be saved in the cache.xml and be persisted in the cluster configuration?

**Comment by mivanac** [comment]
These are existing create gateway-sender options, which now can be modified in runtime. All modified options will be persisted and saved in cluster configuration.

File: geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java:2615]: You should create a new unit test even for inner-classes. Any class that you make Serializable should also have to unit test coverage to ensure that both seriallization and deserialization succeeds and results in a faithful replica of the original object (shamelessly copied from Effective Java 3rd Ed. Item 86). I recommend using org.apache.commons.lang3.SerializationUtils to simplify the unit testing.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java:93]: public mutable fields are not good. I guess this class has other such fields so I'll just skip over this class. If it was me, I'd completely rewrite this class to get make all fields private.

File: geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java]: Removed serialization impacts, so no need for additional tests.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java
Comment (mhansonp) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java:93]: +1

**Review by bschuchardt — COMMENTED** [review]
I am no longer a code-owner for this PR and have removed myself from the list & deleted my now-obsolete comment.

**Review by kirklund — COMMENTED** [review]
The GFSH changes look good to me. I'll add my approval after the WAN changes are reviewed.

**Comment by mivanac** [comment]
HI @kohlmu-pivotal, can you tell me why changes requested, is still on this PR.

**Comment by mivanac** [comment]
HI @kohlmu-pivotal, I see no reason for this CR, because all requests are fullfilled. Could you remove it.

**Comment by kohlmu-pivotal** [comment]
> HI @kohlmu-pivotal, I see no reason for this CR, because all requests are fullfilled. Could you remove it.

Hi there @mivanac, the fate of this PR is in hands of the codeowners. I have nothing to do with this. It is in their hands now.

**Comment by mivanac** [comment]
Hi @jdeppe-pivotal, @jinmeiliao, @mhansonp, @boglesby, @gesterzhou and @nabarunnag, could you review this PR.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java
Comment (albertogpz) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java:103]: What is the purpose of these members and the functions to get/set them? I have not found the use in the rest of the code.

File: geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java
Comment (albertogpz) [inline geode-core/src/main/java/org/apache/geode/cache/configuration/CacheConfig.java:2695]: How about something more compact? return gatewayEventFilters != null;

File: geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java
Comment (albertogpz) [inline geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java:225]: This line is redundant.

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java
Comment (albertogpz) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java:1245]: Is it necessary that the sender is paused while the parameters are changed? If it is so, it would have to be said in the documentation. Otherwise, I would remove the pause here and in the other calls where a gateway sender parameter is changed.

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderAlterOperationsDUnitTest.java
Comment (albertogpz) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderAlterOperationsDUnitTest.java:598]: Same comment as before on the ParallelGateway sender tests. If it is necessary that the gateway sender is paused while parameters are changed, that would have to be documented. Otherwise, I would remove the pause here and in the other parameter changes, below.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderAttributes.java:103]: Thanks for comments. This is not needed, since it was part of previous solution. I will remove it.

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java
Comment (mivanac) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java:1245]: Yes, pausing of gw sender is implemented internally as a part of command execution (see AlterGatewaySenderFunction). Since here we are using only setter, we must ensure that gw sender is paused.

Comment (mivanac) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java:1245]: We pause gw sender to prevent any interference with running processes. This was one of comments on RFC. And I am not sure that internal handling in command execution is documented for users. But if everyone agrees that this is needed, I can add.

**Review by boglesby — COMMENTED** [review]
Setting batch time interval and batch size to invalid values
------------------------------------------------------------
It looks like the minimum for batch size is 1 and batch time interval is -1 (mainly based on looking at code).

Setting batch size to 0 succeeds:
```
(2) Executing - alter gateway-sender --id=ny --batch-size=0

  Member    | Status | Message
----------- | ------ | -------------------------------------------------
server-ln-1 | OK     | GatewaySender ny is updated on member server-ln-1
server-ln-2 | OK     | GatewaySender ny is updated on member server-ln-2
```
But causes this warning:
```
[warn 2021/03/31 11:37:07.105 PDT server-ln-1 <Function Execution Processor2> tid=0x3c] Attempting to set the batch size from 100 to 0 events failed. Instead it was set to 1.
```
And this in the cluster config:
```
<gateway-sender ... batch-size="0" .../>
```
Setting batch time interval to -10 succeeds:
```
(2) Executing - alter gateway-sender --id=ny --batch-time-interval=-10

  Member    | Status | Message
----------- | ------ | -------------------------------------------------
server-ln-1 | OK     | GatewaySender ny is updated on member server-ln-1
server-ln-2 | OK     | GatewaySender ny is updated on member server-ln-2
```
And causes this in cluster.xml
```
<gateway-sender ... batch-time-interval="-10" .../>
```
Setting GatewayEventFilter
--------------------------
Modifying the servers is not atomic. This is not a big deal for most of the parameters, but it could be a problem for GatewayEventFilters. There is potential for one of the members to add a filter before other members.

In that case, the beforeEnqueue method would be called for events in that member and not called for those same events in other members. Depending on the implementation of beforeEnqueue, this would mean an event could be enqueued into the primary queue but not into the secondary (or vice versa).

Pause has no effect on this because beforeEnqueue is invoked in the caller's thread, and pause only affects the processor threads.

Below is some logging that shows this scenario. The *** events are enqueued in one member, but not in the other.

Server ln-2
-----------
```
[info 2021/03/31 13:24:45.814 PDT server-ln-2 <Function Execution Processor2> tid=0x3e] Paused ParallelGatewaySender{id=ny,remoteDsId=2,isRunning =true}

...

***[warn 2021/03/31 13:24:45.866 PDT server-ln-2 <P2P message reader for 192.168.1.15(server-ln-1:37516)<v1>:41001 unshared ordered sender uid=11 dom #1 local port=60428 remote port=49304> tid=0x39] ParallelGatewaySenderEventProcessor.enqueueEvent enqueued event key=437; eventId=EventID[id=25 bytes;threadID=0x10050|2;sequenceID=11271;bucketId=80]

***[warn 2021/03/31 13:24:45.866 PDT server-ln-2 <P2P message reader for 192.168.1.15(server-ln-1:37516)<v1>:41001 unshared ordered sender uid=12 dom #1 local port=60428 remote port=49309> tid=0x24] ParallelGatewaySenderEventProcessor.enqueueEvent enqueued event key=830; eventId=EventID[id=25 bytes;threadID=0x1004b|4;sequenceID=11346;bucketId=75]

***[warn 2021/03/31 13:24:45.866 PDT server-ln-2 <ServerConnection on port 65511 Thread 2> tid=0x5a] ParallelGatewaySenderEventProcessor.enqueueEvent enqueued event key=701; eventId=EventID[id=25 bytes;threadID=0x10027|3;sequenceID=11332;bucketId=39]

...

[warn 2021/03/31 13:24:45.912 PDT server-ln-2 <Function Execution Processor2> tid=0x3e] XXX AbstractGatewaySender.setGatewayEventFilters eventFilters=[TestGatewayEventFilter@3e37dc9f]

...

[info 2021/03/31 13:24:45.913 PDT server-ln-2 <Function Execution Processor2> tid=0x3e] Resumed ParallelGatewaySender{id=ny,remoteDsId=2,isRunning =true}
```
Server ln-1
-----------
```
[info 2021/03/31 13:24:45.813 PDT server-ln-1 <Function Execution Processor2> tid=0x3c] Paused ParallelGatewaySender{id=ny,remoteDsId=2,isRunning =true}

...

[warn 2021/03/31 13:24:45.866 PDT server-ln-1 <Function Execution Processor2> tid=0x3c] XXX AbstractGatewaySender.setGatewayEventFilters eventFilters=[TestGatewayEventFilter@7b1be0cd]

***[warn 2021/03/31 13:24:45.866 PDT server-ln-1 <P2P message reader for 192.168.1.15(server-ln-2:37530)<v2>:41002 unshared ordered sender uid=14 dom #1 local port=43073 remote port=49308> tid=0x3b] ParallelGatewaySenderEventProcessor.enqueueEvent did not enqueue event key=701; eventId=EventID[id=25 bytes;threadID=0x10027|3;sequenceID=11332;bucketId=39]

***[warn 2021/03/31 13:24:45.866 PDT server-ln-1 <ServerConnection on port 65486 Thread 1> tid=0x5d] ParallelGatewaySenderEventProcessor.enqueueEvent did not enqueue event key=437; eventId=EventID[id=25 bytes;threadID=0x10050|2;sequenceID=11271;bucketId=80]

***[warn 2021/03/31 13:24:45.866 PDT server-ln-1 <ServerConnection on port 65486 Thread 2> tid=0x60] ParallelGatewaySenderEventProcessor.enqueueEvent did not enqueue event key=830; eventId=EventID[id=25 bytes;threadID=0x1004b|4;sequenceID=11346;bucketId=75]

[info 2021/03/31 13:24:45.867 PDT server-ln-1 <Function Execution Processor2> tid=0x3c] Resumed ParallelGatewaySender{id=ny,remoteDsId=2,isRunning =true}
```
Rolling Test
------------
If I roll part of the distributed system and attempt to alter the gateway sender, an exception is logged.

1. Start locator and servers in previous version
2. Roll locator and one server to current version
3. alter gateway-sender
```
(2) Executing - alter gateway-sender --id=ny --batch-size=10

  Member    | Status | Message
----------- | ------ | -------------------------------------------------
server-ln-1 | OK     | GatewaySender ny is updated on member server-ln-1
```
The log for server-ln-2 contains this exception:
```
[fatal 2021/03/31 14:18:46.277 PDT <P2P message reader for 10.166.145.32(locator-ln:41902:locator)<ec><v13>:41000(version:VersionOrdinal[ordinal=150]) shared unordered uid=1 local port=45741 remote port=51208> tid=0x61] Error deserializing message
java.lang.ClassNotFoundException: org.apache.geode.management.internal.cli.functions.AlterGatewaySenderFunction
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.Class.forName(Class.java:348)
	at java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:677)
	at org.apache.geode.internal.InternalDataSerializer$DSObjectInputStream.resolveClass(InternalDataSerializer.java:3536)
	at org.apache.geode.internal.InternalDataSerializer.readSerializable(InternalDataSerializer.java:2686)
	at org.apache.geode.internal.InternalDataSerializer.basicReadObject(InternalDataSerializer.java:2630)
	at org.apache.geode.DataSerializer.readObject(DataSerializer.java:2864)
	at org.apache.geode.internal.cache.MemberFunctionStreamingMessage.fromData(MemberFunctionStreamingMessage.java:292)
```
In this case server-ln-1 was rolled and successfully alters the sender, and server-ln-2 was not rolled and did not successfully alter the sender.

I think you can do something what is done in CreateGatewaySenderCommand:
```
if (!verifyAllCurrentVersion(membersToCreateGatewaySenderOn)) {
  return ResultModel.createError(
      CliStrings.CREATE_GATEWAYSENDER__MSG__CAN_NOT_CREATE_DIFFERENT_VERSIONS);
}
```

**Comment by mivanac** [comment]
Thanks for comments. I will update solution.

File: geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java
Comment (jinmeiliao) [inline geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java:73]: A new way to do this is to use `ClassName[]`. The gfsh parser will parse the comma separated list and any initialization json and put them into this ClassName array for you. See `AlterRegionCommand` for examples.

**Review by jinmeiliao — APPROVED** [review]

**Review by gesterzhou — APPROVED** [review]

**Comment by mivanac** [comment]
@ boglesby could we just state that modification of GatewayEventFilter should be performed while there is no traffic,
or should I modify solution to ensure there is no interference.

File: geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java
Comment (mivanac) [reply geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/commands/AlterGatewaySenderCommand.java]: Thanks for comments. Updated.

**Comment by mivanac** [comment]
Since there are no answers, and this PR is open for long time, I will merge it. If you have any comments on this solution, I will open improvement ticket.

**Pull request merged**
